executor/mppcoordmanager/mpp_coordinator_manager.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/copr"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// InstanceMPPCoordinatorManager is a local instance mpp coordinator manager
+var InstanceMPPCoordinatorManager = newMPPCoordinatorManger()
+
+const (
+	detectFrequency = 5 * time.Minute
+)
+
+// CoordinatorUniqueID identifies a unique coordinator
+type CoordinatorUniqueID struct {
+	MPPQueryID kv.MPPQueryID
+	GatherID   uint64
+}
+
+// MPPCoordinatorManager manages all mpp coordinator instances
+type MPPCoordinatorManager struct {
+	mu             sync.Mutex
+	serverOn       bool
+	serverAddr     string // empty if server is off
+	coordinatorMap map[CoordinatorUniqueID]kv.MppCoordinator
+	wg             sync.WaitGroup
+	ctx            context.Context
+	cancel         context.CancelFunc
+	maxLifeTime    uint64 // in Nano
+}
+
+// Run use a loop to detect and remove out of time Coordinators
+func (m *MPPCoordinatorManager) Run() {
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+	m.wg.Add(1)
+	m.maxLifeTime = uint64(copr.TiFlashReadTimeoutUltraLong.Nanoseconds() + detectFrequency.Nanoseconds())
+	go func() {
+		defer m.wg.Done()
+		ticker := time.NewTicker(detectFrequency)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-m.ctx.Done():
+				return
+			case <-ticker.C:
+				m.detectAndDelete(uint64(time.Now().UnixNano()))
+			}
+		}
+	}()
+}
+
+func (m *MPPCoordinatorManager) detectAndDelete(nowTs uint64) {
+	var outOfTimeIDs []CoordinatorUniqueID
+	m.mu.Lock()
+	for id := range m.coordinatorMap {
+		if nowTs > id.MPPQueryID.QueryTs+m.maxLifeTime {
+			outOfTimeIDs = append(outOfTimeIDs, id)
+			delete(m.coordinatorMap, id)
+		}
+	}
+	m.mu.Unlock()
+
+	for _, deletedID := range outOfTimeIDs {
+		metrics.MppCoordinatorStatsOverTimeNumber.Inc()
+		logutil.BgLogger().Error("Delete MppCoordinator due to OutOfTime",
+			zap.Uint64("QueryID", deletedID.MPPQueryID.LocalQueryID),
+			zap.Uint64("QueryTs", deletedID.MPPQueryID.QueryTs))
+	}
+}
+
+// Stop stops background goroutine
+func (m *MPPCoordinatorManager) Stop() {
+	m.cancel()
+	m.wg.Wait()
+}
+
+// InitServerAddr init grpcServer address
+func (m *MPPCoordinatorManager) InitServerAddr(serverOn bool, serverAddr string) {
+	m.serverOn = serverOn
+	if serverOn {
+		m.serverAddr = serverAddr
+	}
+}
+
+// GetServerAddr returns grpcServer address, empty serverAddr if server not on
+func (m *MPPCoordinatorManager) GetServerAddr() (serverOn bool, serverAddr string) {
+	return m.serverOn, m.serverAddr
+}
+
+// Register is to register mpp coordinator
+func (m *MPPCoordinatorManager) Register(coordID CoordinatorUniqueID, mppCoord kv.MppCoordinator) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	_, exists := m.coordinatorMap[coordID]
+	if exists {
+		return errors.Errorf("Mpp coordinator already registered: %d %d %d %d", coordID.MPPQueryID.QueryTs, coordID.MPPQueryID.LocalQueryID, coordID.MPPQueryID.ServerID, coordID.GatherID)
+	}
```

Comment:
from https://github.com/pingcap/tiflash/pull/7694#pullrequestreview-1499880922
Multiple different mpp queries may have the same query id, which may cause the regsiter coodinator to fail?

Reply:
For example, sql with apply

Reply:
I think that may be caused by unused mpp_gather id, in this PR, mpp_gather id will be assigned meaningful value. I'll double check it.

Reply:
Confirmed that, either LocalQueyId, or GatherID is different, different coordinators won't share the same key.

Reply:
ok

Reply:
Do you double checked sql with `Apply` executor? Apply may use the same gather id, if you use `e.id` as the gather id

Reply:
I checked the "select * from t where id > (select max(a) from s)" query, and get the previous conclusion.

Reply:
Maybe concurrent apply is not enabled?

Reply:
https://github.com/pingcap/tidb/pull/44697/files#diff-3fd4cee3c1972379d1b70f86b4cfde79297f1fea3016580771a88dfad9888496R89-R92

Reply:
Yes, record one query that uses apply:
select count(*) from t where a > (select avg(b) from t t1 where t.a > t1.a);